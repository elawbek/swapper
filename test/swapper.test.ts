import { expect } from "chai";
import { ethers } from "hardhat";
import {
  impersonateAccount,
  time,
  loadFixture,
} from "@nomicfoundation/hardhat-network-helpers";
import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";

import { solidityPack, parseEther } from "ethers/lib/utils";

import snapshotGasCost from "./snapshots";
import { WETH, WETH_WAHLE, pairs, tokens, UNI_ROUTER } from "./helper";

import {
  IERC20Metadata__factory,
  IERC20,
  IERC20__factory,
  IRouter,
  IRouter__factory,
  IUniswapV2Pair__factory,
  IUniswapV2Pair,
} from "../typechain-types";

// solc --strict-assembly --optimize --optimize-runs 1000 contracts/Swapper.yul
const bytecode =
  "6101718061004d343933346009015273c02aaa39b223fe8d0a0e5c4f27ead9083c756cc23481816061015260bc01527370997970c51812dc3a010c7d01b50e0d17dc79c83461010a015234f3fe3d341161014957337f000000000000000000000000000000000000000000000000000000000000000014610030575b005b368060351461008b57600e1461004257005b63a9059cbb60e01b3490815233600452803560901c60245280604481807f00000000000000000000000000000000000000000000000000000000000000005af161002e575b3434fd5b5042343560e01c106100875760043560601c63a9059cbb60e01b34528060045260183560901c6024523434604434347f00000000000000000000000000000000000000000000000000000000000000005af1156100875763022c0d9f60e01b345261010534603435116100fd8161014d565b60045261015f565b6024527f0000000000000000000000000000000000000000000000000000000000000000604452608060645234608452349060a43491349034905af161002e573434fd5b3d3dfd5b15610156573490565b60263560901c90565b1561016d5760263560901c90565b349056";

async function deployFixture() {
  const [owner, alice] = await ethers.getSigners();

  await impersonateAccount(WETH_WAHLE);
  const whale = await ethers.getSigner(WETH_WAHLE);

  const weth = IERC20__factory.connect(WETH, owner);
  const token = IERC20__factory.connect(tokens[0], owner);
  const router = IRouter__factory.connect(UNI_ROUTER, owner);
  const pairForTest = IUniswapV2Pair__factory.connect(pairs[0], owner);

  const ContrFactory = new ethers.ContractFactory([], bytecode, owner);
  const swapYulContract = await ContrFactory.deploy();

  return {
    owner,
    alice,
    token,
    weth,
    router,
    pairForTest,
    swapYulContract,
    whale,
  };
}

// hh node
// rm -rf test/__* && hh test test/swapper.test.ts --network localhost
describe("", () => {
  let owner: SignerWithAddress;
  let alice: SignerWithAddress;
  let whale: SignerWithAddress;

  let token: IERC20;
  let weth: IERC20;
  let router: IRouter;
  let pairForTest: IUniswapV2Pair;

  let swapYulContract: any;

  beforeEach(async () => {
    ({
      alice,
      owner,
      weth,
      token,
      router,
      pairForTest,
      swapYulContract,
      whale,
    } = await loadFixture(deployFixture));
  });

  xit("test swap", async () => {
    const amountsOut = await router.getAmountsOut(parseEther("0.1"), [
      WETH,
      token.address,
    ]);

    await weth
      .connect(whale)
      .transfer(swapYulContract.address, parseEther("0.1"));

    const deadline = (await time.latest()) + 1200;

    const data = solidityPack(
      ["uint32", "address", "uint112", "uint112", "bool"],
      [
        deadline,
        pairForTest.address,
        amountsOut[0],
        amountsOut[1],
        (await pairForTest.token0()) === WETH,
      ]
    );

    await expect(() =>
      owner.sendTransaction({
        to: swapYulContract.address,
        data,
        gasLimit: 200000,
        accessList: [
          [
            pairForTest.address,
            [
              "0x0000000000000000000000000000000000000000000000000000000000000006",
              "0x0000000000000000000000000000000000000000000000000000000000000007",
              "0x0000000000000000000000000000000000000000000000000000000000000008",
              "0x0000000000000000000000000000000000000000000000000000000000000009",
              "0x000000000000000000000000000000000000000000000000000000000000000a",
              "0x000000000000000000000000000000000000000000000000000000000000000c",
            ],
          ],
        ],
      })
    ).to.changeTokenBalance(token, alice, amountsOut[1]);
  });

  xit("withdraw test", async () => {
    weth.connect(whale).transfer(swapYulContract.address, 100);

    const withdrawData = solidityPack(["uint112"], [100]);

    await expect(() =>
      owner.sendTransaction({
        to: swapYulContract.address,
        data: withdrawData,
      })
    ).to.changeTokenBalance(weth, owner, 100);
  });

  xit("test gas deploy", async () => {
    const ContrFactory = new ethers.ContractFactory([], bytecode, owner);
    await snapshotGasCost(ContrFactory.deploy());
  });

  it("test gas swaps", async () => {
    await snapshotGasCost(
      weth.connect(whale).transfer(swapYulContract.address, 1)
    );

    for (let i = 0; i < tokens.length; i++) {
      const token = IERC20Metadata__factory.connect(tokens[i], owner);
      const wethMeta = IERC20Metadata__factory.connect(WETH, owner);

      console.log(`Pair: ${await wethMeta.name()} - ${await token.name()}`);
      console.log(`Pair: ${await wethMeta.symbol()} - ${await token.symbol()}`);

      const pair = IUniswapV2Pair__factory.connect(pairs[i], owner);
      console.log(await pair.token1());

      await snapshotGasCost(
        weth.connect(whale).transfer(swapYulContract.address, parseEther("0.1"))
      );

      const deadline = (await time.latest()) + 1200;

      let amountsOut = await router.getAmountsOut(parseEther("0.1"), [
        WETH,
        tokens[i],
      ]);
      const token0 = (await pair.token0()) === weth.address;

      let data = solidityPack(
        ["uint32", "address", "uint112", "uint112", "bool"],
        [deadline, pair.address, amountsOut[0], amountsOut[1], token0]
      );

      await snapshotGasCost(
        owner.sendTransaction({
          to: swapYulContract.address,
          data,
          gasLimit: 200000,
          accessList: [
            [
              pair.address,
              [
                "0x0000000000000000000000000000000000000000000000000000000000000006",
                "0x0000000000000000000000000000000000000000000000000000000000000007",
                "0x0000000000000000000000000000000000000000000000000000000000000008",
                "0x0000000000000000000000000000000000000000000000000000000000000009",
                "0x000000000000000000000000000000000000000000000000000000000000000a",
                "0x000000000000000000000000000000000000000000000000000000000000000c",
              ],
            ],
          ],
        })
      );

      console.log("swap 2\n");

      await snapshotGasCost(
        weth.connect(whale).transfer(swapYulContract.address, parseEther("0.1"))
      );

      amountsOut = await router.getAmountsOut(parseEther("0.1"), [
        WETH,
        tokens[i],
      ]);

      data = solidityPack(
        ["uint32", "address", "uint112", "uint112", "bool"],
        [deadline, pair.address, amountsOut[0], amountsOut[1], token0]
      );

      await snapshotGasCost(
        owner.sendTransaction({
          to: swapYulContract.address,
          data,
          gasLimit: 200000,
          accessList: [
            [
              pair.address,
              [
                "0x0000000000000000000000000000000000000000000000000000000000000006",
                "0x0000000000000000000000000000000000000000000000000000000000000007",
                "0x0000000000000000000000000000000000000000000000000000000000000008",
                "0x0000000000000000000000000000000000000000000000000000000000000009",
                "0x000000000000000000000000000000000000000000000000000000000000000a",
                "0x000000000000000000000000000000000000000000000000000000000000000c",
              ],
            ],
          ],
        })
      );
    }
  }).timeout(2000000);
});
