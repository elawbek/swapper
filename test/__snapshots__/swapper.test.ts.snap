// Jest Snapshot v1, https://goo.gl/fbAQL

exports[` test gas swaps 1`] = `51534`;

exports[` test gas swaps 2`] = `34494`;

exports[` test gas swaps 3`] = `104628`;

exports[` test gas swaps 4`] = `34494`;

exports[` test gas swaps 5`] = `87528`;

exports[` test gas swaps 6`] = `34494`;

exports[` test gas swaps 7`] = `101179`;

exports[` test gas swaps 8`] = `34494`;

exports[` test gas swaps 9`] = `101179`;

exports[` test gas swaps 10`] = `34494`;

exports[` test gas swaps 11`] = `103497`;

exports[` test gas swaps 12`] = `34494`;

exports[` test gas swaps 13`] = `86397`;

exports[` test gas swaps 14`] = `34494`;

exports[` test gas swaps 15`] = `115462`;

exports[` test gas swaps 16`] = `34494`;

exports[` test gas swaps 17`] = `98362`;

exports[` test gas swaps 18`] = `34494`;

exports[` test gas swaps 19`] = `108118`;

exports[` test gas swaps 20`] = `34494`;

exports[` test gas swaps 21`] = `91018`;

exports[` test gas swaps 22`] = `34494`;

exports[` test gas swaps 23`] = `134028`;

exports[` test gas swaps 24`] = `34494`;

exports[` test gas swaps 25`] = `116928`;

exports[` test gas swaps 26`] = `34494`;

exports[` test gas swaps 27`] = `109366`;

exports[` test gas swaps 28`] = `34494`;

exports[` test gas swaps 29`] = `92266`;

exports[` test gas swaps 30`] = `34494`;

exports[` test gas swaps 31`] = `103783`;

exports[` test gas swaps 32`] = `34494`;

exports[` test gas swaps 33`] = `86683`;

exports[` test gas swaps 34`] = `34494`;

exports[` test gas swaps 35`] = `103661`;

exports[` test gas swaps 36`] = `34494`;

exports[` test gas swaps 37`] = `86561`;

exports[` test gas swaps 38`] = `34494`;

exports[` test gas swaps 39`] = `103577`;

exports[` test gas swaps 40`] = `34494`;

exports[` test gas swaps 41`] = `86477`;

exports[` test gas swaps 42`] = `34494`;

exports[` test gas swaps 43`] = `106533`;

exports[` test gas swaps 44`] = `34494`;

exports[` test gas swaps 45`] = `89433`;

exports[` test gas swaps 46`] = `34494`;

exports[` test gas swaps 47`] = `111718`;

exports[` test gas swaps 48`] = `34494`;

exports[` test gas swaps 49`] = `94618`;

exports[` test gas swaps 50`] = `34494`;

exports[` test gas swaps 51`] = `103564`;

exports[` test gas swaps 52`] = `34494`;

exports[` test gas swaps 53`] = `86464`;

exports[` test gas swaps 54`] = `34494`;

exports[` test gas swaps 55`] = `108117`;

exports[` test gas swaps 56`] = `34494`;

exports[` test gas swaps 57`] = `91017`;

exports[` test gas swaps 58`] = `34494`;

exports[` test gas swaps 59`] = `103699`;

exports[` test gas swaps 60`] = `34494`;

exports[` test gas swaps 61`] = `86599`;

exports[` test gas swaps 62`] = `34494`;

exports[` test gas swaps 63`] = `103610`;

exports[` test gas swaps 64`] = `34494`;

exports[` test gas swaps 65`] = `86510`;

exports[` test gas swaps 66`] = `34494`;

exports[` test gas swaps 67`] = `181580`;

exports[` test gas swaps 68`] = `34494`;

exports[` test gas swaps 69`] = `164468`;

exports[` test gas swaps 70`] = `34494`;

exports[` test gas swaps 71`] = `110535`;

exports[` test gas swaps 72`] = `34494`;

exports[` test gas swaps 73`] = `93435`;

exports[` test gas swaps 74`] = `34494`;

exports[` test gas swaps 75`] = `103523`;

exports[` test gas swaps 76`] = `34494`;

exports[` test gas swaps 77`] = `86423`;

exports[` test gas swaps 78`] = `34494`;

exports[` test gas swaps 79`] = `113875`;

exports[` test gas swaps 80`] = `34494`;

exports[` test gas swaps 81`] = `96775`;
